### PR TITLE
fix(factual_research): raise max_tokens, return error JSON, reduce overconfidence

### PR DESCRIPTION
**Pull request by LOCKhart07**
## Context

Post-launch analysis of `factual_research` on Gnosis over 2026-04-15 → 2026-04-16 (209 delivers, tool went live 2026-04-15) surfaced two problems:

- **~20% of deliveries returned raw framework error strings.** Every failing call had `completion_tokens=1500` — the model was hitting the output cap mid-JSON, raising `openai.LengthFinishReasonError`, which wasn't caught in `_parse_completion`'s retry block and bubbled up as `str(e)` through `with_key_rotation`.
- **Output distribution was heavily tail-loaded.** 85.7% of structured answers at `p_yes<0.1` or `>0.9`, only 2.4% in the genuinely uncertain middle band (`0.3–0.5`), mean confidence 0.984 in the `p_yes>0.9` bucket. No ground-truth calibration available yet (markets haven't resolved), but the shape alone was out of line with sibling tools.

## Changes

Three focused edits in `packages/valory/customs/factual_research/factual_research.py`:

**Raise `max_tokens` 1500 → 4096** on both the synthesis and estimation LLM calls. Expected to drop the non-JSON delivery rate from ~20% to near zero.

**`with_key_rotation` error path returns a well-formed error JSON** — `{p_yes:null, p_no:null, confidence:0.0, info_utility:0.0, error:<msg>}` — instead of stringifying the exception. Downstream consumers can now parse-and-branch on failures rather than receiving framework text.

**Estimate prompt tightened against overconfidence:**
- Symmetric tail discipline. The prior version only required justification for `p_yes > 0.9`; the low tail had no cap, which matched the 81% concentration at `p_yes < 0.1` in prod.
- Softer "absence of expected evidence = NO". Now conditional on timeline — for future-dated events where the deadline is still weeks/months out, absence of recent news is neutral, not negative.
- Base-rate stickiness. Weak/procedural signals may move at most ±0.10; extreme tails (`<0.1` or `>0.9`) now explicitly require **multiple independent strong signals**.

Plus the lockfile regeneration commit (`autonomy packages lock`) — the factual_research custom hash and chained agent/service hashes, needed to pass `linter_checks` CI.

## Not in this PR

- Synthesis-prompt changes and anchored-output schema (report suggestion 3). Deferred — see "Next levers" below.
- Alerting on `non-JSON toolResponse rate > 2%` (report suggestion 4). Ops/monitoring, not in this repo.
- Calibration curve / Brier score (report suggestion 6). Needs market resolutions, re-run in 2–3 weeks.

## Validation — distribution replay

An ad-hoc script (not included here) pulled 30 recent `factual_research` deliveries from the Gnosis subgraph, extracted their cached `source_content` from IPFS, and re-ran the tool locally against the new prompt. Outcome-free distribution comparison:

```
bucket       baseline      candidate     delta
----------------------------------------------
<0.1         24 (80.0%)    23 (76.7%)    -3.3%
0.1-0.3       1 ( 3.3%)     2 ( 6.7%)    +3.3%
0.3-0.5       1 ( 3.3%)     0 ( 0.0%)    -3.3%
0.5-0.7       1 ( 3.3%)     2 ( 6.7%)    +3.3%
0.7-0.9       2 ( 6.7%)     2 ( 6.7%)    +0.0%
>=0.9         1 ( 3.3%)     1 ( 3.3%)    +0.0%

mean    0.146 → 0.156
median  0.025 → 0.035
tails   83.3% → 80.0%
```

Direction is correct (mean/median both up, tail concentration down, one `0.12 → 0.55` recovery into the uncertain band, one `0.85 → 0.01` reversal where the briefing contained strong negative evidence) but the shift is modest at the estimate-prompt layer alone. 23 of 24 rows that were `p_yes < 0.1` on baseline stayed `p_yes < 0.1` on candidate — the factual briefing upstream is the dominant signal, and the estimate prompt can only nudge from there.

## Validation — 10-row correctness audit

The distribution replay answers "does the shape of p_yes change?" — it does not answer "are the extreme outputs actually wrong?" To check the latter, pulled the **full prod stage trace** (reframe → sub-questions → evidence → synthesis → briefing → estimate → reasoning) from IPFS for 10 baseline rows with p_yes < 0.05 or > 0.95, via the `prompt` field of each delivery's IPFS payload. For each row, cross-checked whether the briefing contained evidence that mathematically justified the extreme, or whether the extreme was an artifact of the "absence = NO" shortcut.

| n | base → cand | evidence basis | verdict |
|---|---|---|---|
| 1 | 0.995 → 0.99  | US-Iran peace talks already publicly confirmed as begun | concrete positive evidence |
| 2 | 0.030 → 0.12  | Hancock Prospecting royalty — separate trial pending for payout amount | procedural barrier |
| 3 | 0.010 → 0.01  | Air NZ Skynest — service scheduled Nov 2026, question deadline Apr 20 | explicit disqualification |
| 4 | 0.040 → 0.07  | Intel/Google joint IPU — base rate + 4-day window | defensible via base rate |
| 5 | 0.030 → 0.03  | CA/NY AG probe — base rate + 4-day window | defensible via base rate |
| 6 | 0.010 → 0.03  | Mercedes EQS — MB's own EPA estimate 425mi < 500mi question threshold | specific negative evidence |
| 7 | 0.010 → 0.02  | Xbox Game Pass price cut — zero historical precedent | strong historical base rate |
| 8 | 0.010 → 0.02  | WB/IMF reforms — deadline is "today", reforms take years | deadline effectively missed |
| 9 | 0.001 → 0.01  | Duplicate of row 3 | same explicit disqualification |
| 10| 0.010 → 0.03  | LaLiga/Fastly expansion — base rate + 4-day window | defensible via base rate |

**Verdict: 10/10 extreme outputs are evidence-justified.** Breakdown:
- 6/10 via concrete evidence (explicit disqualification, specific negative evidence, or event already happened).
- 3/10 via (1-5% base rate) + (4-day deadline window) — output is base-rate-driven, not absence-as-NO-driven.
- 1/10 via procedural barrier (separate trial required).

### Why the distribution shift is small (secondary finding)

7/10 baseline reasonings explicitly invoke "absence of expected evidence is a strong NO signal" — the exact pattern this PR targets. **But** on rows where the baseline reasoning would have reached the same conclusion via base rate alone (short deadlines, rare event types), the softened prompt can't move the output much. The 30-row sample is skewed toward short-deadline "will X publicly announce Y in N days" questions where base rates legitimately sit at 1-5%.

The PR's measurable wins are on rows where absence-as-NO was the decisive driver:
- `0.12 → 0.55` (UK Gambling Commission): briefing was ambiguous, old prompt over-weighted NO
- `0.85 → 0.01` (US vehicle avg transaction price): briefing contained strong negative evidence that old prompt under-weighted

These rows exist but are a minority in the sampled 30. Ship and re-evaluate once markets resolve.

## Validation — token-cap fix (two independent lines of evidence)

The initial "~20% error rate" number came from a 48h snapshot. Expanded to a 7-day window (2026-04-14 → 2026-04-21) to confirm: **1190 factual_research deliveries, 290 failures = 24.4%**. Then verified the fix two ways.

### 1. Deterministic: 100% of failures are the same root cause

Parsing the `toolResponse` string of all 290 failing rows: **every single one** emits the `openai.LengthFinishReasonError` text, and the `CompletionUsage(completion_tokens=1500, ...)` field reports `completion_tokens=1500` exactly (= the old cap). Prompt tokens range ~3800-3900. The failure mode is the structured-output JSON getting truncated mid-generation because the model's reasoning budget hit the cap.

Raising the cap to 4096 gives the model 2.7x the token budget. Any row that previously hit 1500 now has room to finish.

Residual risk is the tiny edge case where the model wants >4096 tokens. The error-JSON fallback in `with_key_rotation` catches those cleanly, so no raw framework strings leak to consumers.

### 2. Empirical: 10/10 fresh-run pass on prod-failed markets

Randomly sampled 10 of the 290 failing markets (seed=42) and re-ran the tool **fresh** (live Serper search + web scrape, no cached `source_content` — we couldn't cache-replay because failing rows never persist `source_content` to IPFS since the crash happens inside the estimate step, before the IPFS write).

| n | status | p_yes | conf | question |
|---|---|---|---|---|
| 1 | valid | 0.90 | 0.8  | Strait of Hormuz closed to Iranian-linked ships |
| 2 | valid | 0.82 | 0.7  | Strait of Hormuz closed to commercial / oil traffic |
| 3 | valid | 0.28 | 0.6  | Strait of Hormuz under Iranian military control |
| 4 | valid | 0.85 | 0.8  | Strait of Hormuz closed to traffic (variant) |
| 5 | valid | 0.25 | 0.6  | Strait of Hormuz closed to traffic (variant) |
| 6 | valid | 0.65 | 0.6  | Strait of Hormuz open for free passage |
| 7 | valid | 0.32 | 0.55 | Strait of Hormuz under Iranian military control (variant) |
| 8 | valid | 0.001| 0.99 | Big Falls Dam structural failure |
| 9 | valid | 0.88 | 0.8  | OECD maintains [...] |
| 10| valid | 0.03 | 0.9  | Company / institution announces semiconductor prototype |

**10/10 returned parseable structured JSON.** Latencies 24-39s, no crashes, no unparseable responses, no error-JSON triggered (all `p_yes` non-null). Middle-band occupancy is notable: 4/10 rows (rows 3, 5, 6, 7 — each asking about the Strait of Hormuz at different wording angles) land at 0.25-0.65, which is the kind of "I'm genuinely uncertain about a live geopolitical situation" the new prompt is supposed to permit.

### Projected post-merge failure rate

If 4096 captures ~99% of what the model wants to write (conservative given the 2.7x headroom over the distribution that was hitting 1500), the non-JSON `toolResponse` rate should drop from **24.4% → <1%**. The remaining <1% falls back to the parseable error JSON rather than raw framework text.

## Review follow-up (commit `2e1a96e8`)

Two review threads on the error-JSON branch addressed in a follow-up commit:

**Auth/permission errors now rotate the key** (per @bennyjo's option (a)). `with_key_rotation` previously caught `openai.RateLimitError` in the rotate branch, letting `AuthenticationError` / `PermissionDeniedError` fall through to the generic `except Exception` wrap — post-PR they were indistinguishable at the response level from a `LengthFinishReasonError`. The rotate branch now catches all three; a revoked key gets rotated out, same as a throttled one. The `raise` path on exhausted retries is preserved so the framework still sees "pool is burned" distinctly.

**Error-JSON contract pinned by dedicated tests** (per @bennyjo's first inline + @DIvyaNautiyal07's P2). The original tests used `assert "..." in result[0]` — substring-matching that would still pass on a regression to `return str(e), ...`. Tightened:

- **`TestRunEdgeCases`** (3 existing tests): now share a `_assert_error_envelope` helper that parses the JSON and asserts `p_yes is None`, zeroed scores, and substring in `error`. Regressing the decorator now fails at `json.loads`.
- **`TestWithKeyRotation`** (new class, 8 cases): success path tuple shape, max-cost float pass-through, rotate-on-recoverable-error (parametrized over `RateLimit` / `Authentication` / `PermissionDenied`), exhausted-retries re-raises, unhandled-exception wraps to error JSON, and an end-to-end `@patch("_parse_completion")` + `RuntimeError("simulated truncation")` regression matching the literal production bug shape.

Stale docstring on `TestRunEdgeCases` also updated to describe the JSON-wrapped contract.

## Next levers (if the modest shift isn't enough post-merge)

1. Weaken the synthesis prompt so the briefing itself doesn't frame "absence" as negative evidence on future-dated questions.
2. Schema-level change: anchored probability enum `{0.05, 0.2, 0.35, 0.5, 0.65, 0.8, 0.95}` — guarantees no sub-0.05 extremes.
3. Re-run the replay at n≥100 to narrow the confidence interval on the distribution shift.

## Test plan

- [ ] CI passes (unit tests in `packages/valory/customs/factual_research/tests/` — all 29 unchanged tests still green locally).
- [ ] `tomte check-code` clean (verified locally).
- [ ] `tox -e check-hash` clean (verified locally post-lock).
- [x] Pre-merge: token-cap fix verified against 290 prod-failed rows (100% root-cause confirmed) + 10/10 fresh-run pass.
- [x] Review follow-up (`2e1a96e8`): auth-error rotation + `TestWithKeyRotation` class pinning the decorator contract.
- [ ] Post-merge: monitor non-JSON `toolResponse` rate over a rolling 1h window. Expect <1% within a day of deployment.
- [ ] Post-merge: verify the distribution shift replicates on fresh deliveries (the replay script is local to my machine; can land it as a follow-up if useful).

### CODE REVIEW

**Review by jmoreira-valory — DISMISSED** [review]

**Review by OjusWiZard — DISMISSED** [review]

**Comment by LOCKhart07** [comment]
Integration tests are failing because Anthropic retired the claude model `claude-3-haiku-20240307` yesterday https://platform.claude.com/docs/en/about-claude/model-deprecations
Needs to be fixed but out of scope for this pr

**Review by DIvyaNautiyal07 — COMMENTED** [review]
Two non-blocking notes, both on the error-JSON change in `with_key_rotation`. Core code changes (max_tokens 1500→4096, prompt edits, error-JSON wrapping) look solid. Validation in the PR body is strong — 290 prod-failed rows confirmed same root cause, 10/10 fresh-run pass.

**P2 — no test pins the new error-JSON shape**
`tests/test_factual_research.py` isn't in this PR, but the existing tests in `TestRunEdgeCases` (`test_rejects_unknown_tool`, `test_rejects_missing_model`, `test_invalid_source_content_mode`) use `assert "..." in result[0]`, which is substring-matching against the full JSON string. They would still pass if someone regressed the decorator back to `return str(e), ...`. Consider adding one assertion like:
```python
data = json.loads(result[0])
assert data["p_yes"] is None and "not supported" in data["error"]
```
on one of them, to lock in the contract described in the PR body ("parseable failure markers instead of garbage text").

**Nit — stale docstring in the same test class**
`TestRunEdgeCases` docstring (around line 650 on main) says "returns an error tuple `(error_str, "", None, None, None, api_keys)`". After this PR the first element is an error JSON, not a raw `error_str`. Worth updating to match.

**Review by DIvyaNautiyal07 — DISMISSED** [review]

File: packages/valory/customs/factual_research/factual_research.py [hunk 1]
```diff
@@ -234,7 +234,16 @@ def execute() -> Union[MaxCostResponse, MechResponseWithKeys]:
...
+                error_json = json.dumps(
+                    {
+                        "p_yes": None,
+                        "p_no": None,
+                        "confidence": 0.0,
+                        "info_utility": 0.0,
+                        "error": str(e),
+                    }
+                )
+                return error_json, "", None, None, None, api_keys
```
Comment (bennyjo) [inline packages/valory/customs/factual_research/factual_research.py:253]: This new error-JSON branch doesn't have a direct test. Two gaps worth closing:

1. **`TestRunEdgeCases` is non-discriminating.** `assert "not supported" in result[0]` (test_factual_research.py:662) and its siblings match both the old `str(e)` form *and* the new JSON-wrapped form, so the contract added here isn't actually validated by the suite. Tightening at least one test to:
   ```python
   parsed = json.loads(result[0])
   assert parsed["p_yes"] is None
   assert parsed["confidence"] == 0.0
   assert "not supported" in parsed["error"]
   ```
   would lock it in.

2. **No regression test for the literal bug.** `_parse_completion`'s retry whitelist (lines 528–533) intentionally doesn't catch `LengthFinishReasonError`, which is why truncation reaches this branch. A small test would protect against a refactor that re-routes the error back to raw-string form:
   ```python
   @patch(f"{FR_MODULE}._parse_completion")
   def test_unhandled_exception_returns_parseable_error_json(self, mock_parse):
       mock_parse.side_effect = RuntimeError("simulated truncation")
       result = run(tool="factual_research", model="gpt-4.1-2025-04-14",
                    prompt=PREDICTION_PROMPT, api_keys=_make_mock_api_keys())
       parsed = json.loads(result[0])
       assert parsed["p_yes"] is None
       assert "simulated truncation" in parsed["error"]
   ```

**Review by bennyjo — COMMENTED** [review]
New error-JSON contract has no discriminating test coverage — two small additions would lock it in.

File: packages/valory/customs/factual_research/factual_research.py (hunk 1 above)
Comment (bennyjo) [inline packages/valory/customs/factual_research/factual_research.py:253]: One adjacent behavior worth confirming — this branch now also catches `openai.AuthenticationError` / `PermissionDeniedError` and wraps them in the same error JSON. Pre-PR these surfaced as raw `"Invalid API key provided: ..."` strings, which is loud; post-PR they're indistinguishable from a `LengthFinishReasonError` at the response level unless you parse the `error` field.

Since the decorator is named `with_key_rotation`, was it intentional that auth/permission failures don't rotate the key like `RateLimitError` does? A revoked key in the rotation pool will produce clean error JSON for every call that hits it, until someone notices via the `error` field. Either:
- add `openai.AuthenticationError` / `PermissionDeniedError` to the rotation branch alongside `RateLimitError`, or
- let them propagate (remove from the `except Exception` sweep) so the mech framework sees the auth failure distinctly.

Happy to leave as-is if this is deliberate — just flagging that the new error contract hides a signal that used to be loud.

**Review by bennyjo — COMMENTED** [review]
Follow-up: the new error-JSON sweep now also covers auth-failure exceptions — worth confirming that's intended.

File: packages/valory/customs/factual_research/factual_research.py (hunk 1 above)
Comment (LOCKhart07) [reply packages/valory/customs/factual_research/factual_research.py:253]: Addressed in 2e1a96e8.

For (1), tightened `TestRunEdgeCases` via a shared `_assert_error_envelope` helper — all three edge-case tests now parse the first tuple element and check the full envelope shape (`p_yes is None`, zeroed scores, substring in `error`). A regression to `return str(e), ...` would now fail at `json.loads`.

For (2), added `TestWithKeyRotation::test_run_wrapped_unhandled_exception_returns_error_json` that uses the `@patch(f"{FR_MODULE}._parse_completion")` + `RuntimeError("simulated truncation")` pattern you outlined — it's the exact literal-bug regression test. Also added a simpler direct-decorator test `test_unhandled_exception_returns_parseable_error_json` that exercises the same contract without the `run()` pipeline in the way.

Comment (LOCKhart07) [reply packages/valory/customs/factual_research/factual_research.py:253]: Good catch — went with option (a). In 2e1a96e8 the `except openai.RateLimitError` clause is extended to:

```python
except (
    openai.RateLimitError,
    openai.AuthenticationError,
    openai.PermissionDeniedError,
) as e:
```

with an inline comment explaining the rationale (a revoked key is as unusable as a throttled one, and rotating it out is the decorator's whole job).

Tests: `TestWithKeyRotation::test_rotates_on_recoverable_error` is parametrized over all three exception classes, and `test_rotation_exhausted_raises` confirms that when the pool is burned, the exception propagates rather than being silently wrapped.

**Comment by LOCKhart07** [comment]
@DIvyaNautiyal07 thanks for the review — both points addressed in 2e1a96e8.

> `TestRunEdgeCases` uses `assert "..." in result[0]`, which is substring-matching against the full JSON string. They would still pass if someone regressed the decorator back to `return str(e), ...`. Consider adding one assertion like: `data = json.loads(result[0]); assert data["p_yes"] is None and "not supported" in data["error"]`.

Tightened the three existing edge-case tests via a shared `_assert_error_envelope` helper that does exactly that (parse first, then check `p_yes is None`, zeroed scores, and the substring in `error`). A regression to `return str(e), ...` now fails at `json.loads`.

> **Nit — stale docstring in the same test class**
> `TestRunEdgeCases` docstring says "returns an error tuple `(error_str, "", None, None, None, api_keys)`". After this PR the first element is an error JSON, not a raw `error_str`.

Updated — docstring now describes the JSON-wrapped contract explicitly.

Also picked up two adjacent items from @bennyjo's inline threads: rotate on `AuthenticationError` / `PermissionDeniedError` (since the decorator's whole point is rotating bad keys), and a new `TestWithKeyRotation` class pinning the decorator directly (success / max-cost float / rotate-on-recoverable-error parametrized / exhausted-raises / unhandled-wraps-to-error-JSON / end-to-end `_parse_completion` side-effect).

**Review by bennyjo — DISMISSED** [review]
Follow-up addresses both threads — rotate branch now covers auth/permission errors and the error-JSON contract is pinned by discriminating tests. LGTM.

**Comment by LOCKhart07** [comment]
Will rebase off of main after https://github.com/valory-xyz/mech-predict/pull/239 lands

**Comment by LOCKhart07** [comment]
/benchmark factual_research --sample 10 --seed 67

^ Did not work because factual research is not properly integrated in prompt replay. Fixing in another pr

**Review by OjusWiZard — APPROVED** [review]

**Review by jmoreira-valory — APPROVED** [review]

**Review by bennyjo — APPROVED** [review]
Re-review LGTM. Code unchanged since prior approval on 2e1a96e — only merge-from-main + hash regen in 240f210 / de66365. CI green on lock_check / linter_checks / copyright_and_dependencies_check / gitleaks / all 9 unit_tests matrix cells; integration_tests still running.

File: packages/valory/customs/factual_research/tests/test_factual_research.py [hunk 2]
```diff
@@ -699,3 +718,145 @@ def test_delivery_rate_zero_returns_max_cost(self) -> None:
...
+        """Rate-limit, auth, and permission errors all rotate the key and retry."""
+        keys = _make_mock_api_keys()
+        keys.max_retries = lambda: {"openai": 1, "openrouter": 1}
+        keys.rotate = MagicMock()
+        call_count = {"n": 0}
+
+        @module.with_key_rotation
+        def fake(api_keys: Any) -> Tuple[Any, ...]:  # pylint: disable=unused-argument
+            call_count["n"] += 1
+            if call_count["n"] == 1:
```
Comment (DIvyaNautiyal07) [inline packages/valory/customs/factual_research/tests/test_factual_research.py:793]: Nit: counter-based dispatch (`call_count["n"] += 1; if call_count["n"] == 1`) is the same pattern this file avoids in `_make_mock_parse_completion` (which uses a `side_effect` list). Consider:

```python
inner = MagicMock(side_effect=[_make_openai_error(exc_cls, "simulated"), ("ok", "", None, None, None)])
fake = module.with_key_rotation(inner)
```

Fails loudly if the retry loop changes shape, instead of silently reordering under a counter.

File: packages/valory/customs/factual_research/tests/test_factual_research.py [hunk 3]
```diff
@@ -699,3 +718,145 @@ def test_delivery_rate_zero_returns_max_cost(self) -> None:
...
+
+    def test_rotation_exhausted_raises(self) -> None:
+        """Recoverable exception is re-raised when retries are exhausted.
+
+        When both openai and openrouter retries are exhausted, the decorator
+        re-raises instead of wrapping into the error JSON — the framework
+        needs the raw exception to know the key pool is burned.
+        """
+        keys = _make_mock_api_keys()
+        keys.max_retries = lambda: {"openai": 0, "openrouter": 0}
```
Comment (DIvyaNautiyal07) [inline packages/valory/customs/factual_research/tests/test_factual_research.py:810]: Nit: `keys.max_retries = lambda: {"openai": 0, "openrouter": 0}` overrides with the exact default that `_make_mock_api_keys()` already returns (line 122). The override is a no-op — can be dropped.

**Review by DIvyaNautiyal07 — APPROVED** [review]

File: packages/valory/customs/factual_research/tests/test_factual_research.py (hunk 2 above)
Comment (LOCKhart07) [reply packages/valory/customs/factual_research/tests/test_factual_research.py:793]: Not fixing as minor and in test file. Fix would require re reviews from all reviewers

File: packages/valory/customs/factual_research/tests/test_factual_research.py (hunk 3 above)
Comment (LOCKhart07) [reply packages/valory/customs/factual_research/tests/test_factual_research.py:810]: Not fixing as minor and in test file. Fix would require re reviews from all reviewers
